File: packages/autonomo-python/autonomo/transport.py
"""
Transport - HTTP server for AI communication

Provides a simple HTTP API that the MCP server can call.
This runs inside the application being tested.
"""

from dataclasses import dataclass
from typing import Callable, Optional, Dict, Any
import json
import os
from http.server import HTTPServer, BaseHTTPRequestHandler
import threading

from .commands import execute_command
from .state import state


def is_dev_mode() -> bool:
    """
    Check if running in development mode.
    
    Returns True if any of:
    - AUTONOMO_DEV env var is set to 'true' or '1'
    - FLASK_DEBUG or FLASK_ENV=development
    - DJANGO_DEBUG is set
    - DEBUG env var is 'true' or '1'
    """
    # Explicit Autonomo flag
    autonomo_dev = os.environ.get("AUTONOMO_DEV", "").lower()
    if autonomo_dev in ("true", "1"):
        return True
    
    # Flask
    if os.environ.get("FLASK_DEBUG") or os.environ.get("FLASK_ENV") == "development":
        return True
    
    # Django
    if os.environ.get("DJANGO_DEBUG"):
        return True
    
    # Generic DEBUG
    debug = os.environ.get("DEBUG", "").lower()
    if debug in ("true", "1"):
        return True
    
    # If none set, assume development (safer for local testing)
    # Set AUTONOMO_DEV=false explicitly to disable in production
    return os.environ.get("AUTONOMO_DEV", "").lower() != "false"


@dataclass
class TransportConfig:
    """Configuration for the Autonomo transport"""
    port: int = 8080
    host: str = "127.0.0.1"
    cors: bool = True
    dev_only: bool = True
    on_start: Optional[Callable[[str], None]] = None
    on_command: Optional[Callable[[str, Optional[str], Optional[str]], None]] = None


class TransportInstance:
    """Running transport instance"""
    
    def __init__(self, url: str, server: HTTPServer, thread: threading.Thread):
        self.url = url
        self._server = server
        self._thread = thread
    
    def stop(self) -> None:
        """Stop the server"""
        self._server.shutdown()
        self._thread.join()


def handle_request(
    method: str,
    path: str,
    body: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    """Handle an incoming HTTP request"""
    import time
    
    # Health check
    if method == "GET" and path == "/health":
        return {
            "status": 200,
            "body": {
                "status": "ok",
                "timestamp": int(time.time() * 1000),
            },
        }
    
    # Get current state
    if method == "GET" and path == "/state":
        return {
            "status": 200,
            "body": state.get_state().to_dict(),
        }
    
    # Execute command
    if method == "POST" and path == "/command":
        if body is None:
            return {
                "status": 400,
                "body": {"error": "Missing request body"},
            }
        
        command = body.get("command")
        target = body.get("target")
        value = body.get("value")
        
        if command is None:
            return {
                "status": 400,
                "body": {"error": "Missing command field"},
            }
        
        result = execute_command(command, target, value)
        return {
            "status": 200 if result.success else 400,
            "body": result.to_dict(),
        }
    
    # Not found
    return {
        "status": 404,
        "body": {"error": "Not found"},
    }


def create_http_transport(config: TransportConfig) -> Optional[TransportInstance]:
    """
    Create and start HTTP transport.
    
    Returns None if dev_only=True and not in development mode.
    """
    # Skip in production if dev_only is enabled
    if config.dev_only and not is_dev_mode():
        return None
    
    class RequestHandler(BaseHTTPRequestHandler):
        def _set_cors_headers(self):
            if config.cors:
                self.send_header("Access-Control-Allow-Origin", "*")
                self.send_header("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
                self.send_header("Access-Control-Allow-Headers", "Content-Type")
        
        def do_OPTIONS(self):
            self.send_response(200)
            self._set_cors_headers()
            self.end_headers()
        
        def do_GET(self):
            result = handle_request("GET", self.path)
            self.send_response(result["status"])
            self.send_header("Content-Type", "application/json")
            self._set_cors_headers()
            self.end_headers()
            self.wfile.write(json.dumps(result["body"]).encode())
        
        def do_POST(self):
            content_length = int(self.headers.get("Content-Length", 0))
            body = None
            if content_length > 0:
                body = json.loads(self.rfile.read(content_length).decode())
            
            result = handle_request("POST", self.path, body)
            self.send_response(result["status"])
            self.send_header("Content-Type", "application/json")
            self._set_cors_headers()
            self.end_headers()
            self.wfile.write(json.dumps(result["body"]).encode())
        
        def log_message(self, format, *args):
            pass  # Suppress logging
    
    server = HTTPServer((config.host, config.port), RequestHandler)
    url = f"http://{config.host}:{config.port}"
    
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()
    
    if config.on_start:
        config.on_start(url)
    
    return TransportInstance(url, server, thread)

File: packages/autonomo-ruby/lib/autonomo/transport.rb
# frozen_string_literal: true

require "webrick"
require "json"

module Autonomo
  # Configuration for the Autonomo transport
  class TransportConfig
    attr_accessor :port, :host, :cors, :on_start, :on_command

    def initialize(port: 8080, host: "127.0.0.1", cors: true, on_start: nil, on_command: nil)
      @port = port
      @host = host
      @cors = cors
      @on_start = on_start
      @on_command = on_command
    end
  end

  # Running transport instance
  class TransportInstance
    attr_reader :url

    def initialize(url, server, thread)
      @url = url
      @server = server
      @thread = thread
    end

    def stop
      @server.shutdown
      @thread.join
    end
  end

  # HTTP transport for AI communication
  module Transport
    class << self
      # Handle an incoming HTTP request
      def handle_request(method, path, body = nil)
        # Health check
        if method == "GET" && path == "/health"
          return {
            status: 200,
            body: { status: "ok", timestamp: (Time.now.to_f * 1000).to_i }
          }
        end

        # Get current state
        if method == "GET" && path == "/state"
          return {
            status: 200,
            body: Autonomo.state.get_state.to_h
          }
        end

        # Execute command
        if method == "POST" && path == "/command"
          unless body
            return {
              status: 400,
              body: { error: "Missing request body" }
            }
          end

          command = body["command"]
          target = body["target"]
          value = body["value"]

          unless command
            return {
              status: 400,
              body: { error: "Missing command field" }
            }
          end

          result = Commands.execute(command, target, value)
          return {
            status: result.success ? 200 : 400,
            body: result.to_h
          }
        end

        # Not found
        {
          status: 404,
          body: { error: "Not found" }
        }
      end

      # Create and start HTTP transport
      def create_http_transport(config = TransportConfig.new)
        url = "http://#{config.host}:#{config.port}"

        server = WEBrick::HTTPServer.new(
          Port: config.port,
          BindAddress: config.host,
          Logger: WEBrick::Log.new("/dev/null"),
          AccessLog: []
        )

        # CORS and request handling
        server.mount_proc "/" do |req, res|
          # CORS headers
          if config.cors
            res["Access-Control-Allow-Origin"] = "*"
            res["Access-Control-Allow-Methods"] = "GET, POST, OPTIONS"
            res["Access-Control-Allow-Headers"] = "Content-Type"
          end

          # Handle OPTIONS
          if req.request_method == "OPTIONS"
            res.status = 200
            next
          end

          # Parse body
          body = nil
          if req.body && !req.body.empty?
            body = JSON.parse(req.body)
          end

          # Handle request
          result = handle_request(req.request_method, req.path, body)

          res.status = result[:status]
          res["Content-Type"] = "application/json"
          res.body = JSON.generate(result[:body])
        end

        thread = Thread.new { server.start }

        config.on_start&.call(url)

        TransportInstance.new(url, server, thread)
      end
    end
  end

  # Module-level helpers
  class << self
    def create_http_transport(config = TransportConfig.new)
      Transport.create_http_transport(config)
    end

    def handle_request(method, path, body = nil)
      Transport.handle_request(method, path, body)
    end
  end
end
